File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.go
func maxProfit(prices []int) int {
    minPrice, profit := prices[0], 0
    
    for _, price := range prices {
        if price < minPrice {
            minPrice = price
        } else if newProfit := price - minPrice; profit < newProfit {
            profit = newProfit
        }
    }
    return profit
}

File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.ts
// Single Pass
function maxProfit(prices: number[]): number {
  let minPrice = Infinity;
  let profit = 0;
  
  prices.forEach(function (price) {
    if (price < minPrice) {
      minPrice = price;
    } else {
      profit = Math.max(profit, price - minPrice);
    }
  });
  
  return profit;
};